x/gamm/keeper/estimate_swap.go

```diff
@@ -0,0 +1,200 @@
+	ctx sdk.Context,
+	routes []types.SwapAmountInRoute,
+	tokenIn sdk.Coin,
+	tokenOutMinAmount sdk.Int,
+) (tokenOutAmount sdk.Int, err error) {
+	// use cache context so that pool state is not mutated after estimation
+	cacheCtx, _ := ctx.CacheContext()
+	for i, route := range routes {
+		swapFeeMultiplier := sdk.OneDec()
+		if types.SwapAmountInRoutes(routes).IsOsmoRoutedMultihop() {
+			swapFeeMultiplier = types.MultihopSwapFeeMultiplierForOsmoPools.Clone()
+		}
+
+		// To prevent the multihop swap from being interrupted prematurely, we keep
+		// the minimum expected output at a very low number until the last pool
+		_outMinAmount := sdk.NewInt(1)
+		if len(routes)-1 == i {
+			_outMinAmount = tokenOutMinAmount
+		}
+
+		// Execute the expected swap on the current routed pool
+		pool, poolErr := k.getPoolForSwap(cacheCtx, route.PoolId)
+		if poolErr != nil {
+			return sdk.Int{}, poolErr
+		}
+
+		swapFee := pool.GetSwapFee(cacheCtx).Mul(swapFeeMultiplier)
+
+		// Execute the expected swap on the current routed pool
+		// call internal method with useCacheCtx = false to prevent double caching,
+		// we need to save pool state for the next iteration of pool route
+		tokenOutAmount, err = k.estimateSwapExactAmountIn(cacheCtx, route.PoolId, tokenIn, route.TokenOutDenom, _outMinAmount, swapFee, false)
+		if err != nil {
+			return sdk.Int{}, err
+		}
+
+		// Chain output of current pool as the input for the next routed pool
+		tokenIn = sdk.NewCoin(route.TokenOutDenom, tokenOutAmount)
+	}
+	return tokenOutAmount, nil
+}
+
+// EstimateSwapExactAmountIn estimates the amount of token out given the exact amount of token in.
+// This method does not execute the full steps of an actaul swap, nor does it change the state after estimation.
+// Returns the estimated amount of token out using cache context.
+func (k Keeper) EstimateSwapExactAmountIn(
+	ctx sdk.Context,
+	poolId uint64,
+	tokenIn sdk.Coin,
+	tokenOutDenom string,
+	tokenOutMinAmount sdk.Int,
+	swapFee sdk.Dec,
+) (sdk.Int, error) {
+	return k.estimateSwapExactAmountIn(ctx, poolId, tokenIn, tokenOutDenom, tokenOutMinAmount, swapFee, true)
+}
+
+// estimateSwapExactAmountIn returns the amount of token out given the amount of token in.
+// This method should only be called by estimate methods, as this method does not include full steps of an actual swap.
+func (k Keeper) estimateSwapExactAmountIn(
+	ctx sdk.Context,
+	poolId uint64,
+	tokenIn sdk.Coin,
+	tokenOutDenom string,
+	tokenOutMinAmount sdk.Int,
+	swapFee sdk.Dec,
+	useCacheCtx bool,
+) (sdk.Int, error) {
+	if useCacheCtx {
+		ctx, _ = ctx.CacheContext()
+	}
+	pool, err := k.getPoolForSwap(ctx, poolId)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	tokenOut, err := k.swapExactAmountInNoTokenSend(ctx, pool, tokenIn, tokenOutDenom, tokenOutMinAmount, swapFee)
+	if err != nil {
+		return tokenOut.Amount, err
+	}
+
+	err = k.updatePoolForSwap(ctx, pool, tokenIn, tokenOut)
+	return tokenOut.Amount, err
+}
```

Comment:
Can we make it clearer what the diff is between these two? Both say "this method does not include full steps of an actual swap", but it seems like if cache context is false then we actually mutate state

Reply:
Ahh, yeah this comment does need some better explanation. Added some more in line comments on this, tell me if it makes more sense now!